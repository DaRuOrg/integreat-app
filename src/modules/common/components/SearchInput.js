import React from 'react'
import PropTypes from 'prop-types'
import FontAwesome from 'react-fontawesome'
import { translate } from 'react-i18next'

import style from './SearchInput.css'

class SearchInput extends React.Component {
  static propTypes = {
    filterText: PropTypes.string.isRequired,
    onFilterTextChange: PropTypes.func.isRequired,
    spaceSearch: PropTypes.bool
  }

  render () {
    const {t} = this.props
    return (
      <div className={this.props.spaceSearch ? style.searchSpacing : undefined}>
        <div className={style.search}>
          <FontAwesome className={style.searchIcon} name='search' />
          <input type='text' placeholder={t('search')} className={style.searchInput} defaultValue={this.props.filterText}
                 onChange={(event) => this.props.onFilterTextChange(event.target.value)} />
        </div>
      </div>
    )
  }
}

export default translate('common')(SearchInput)
