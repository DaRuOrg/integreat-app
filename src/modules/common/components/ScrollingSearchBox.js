import React from 'react'
import PropTypes from 'prop-types'

import { animateScroll } from 'react-scroll'
import SearchInput from './SearchInput'
import StickyHeadroom from './StickyHeadroom'

const SCROLL_ANIMATION_DURATION = 500

export class ScrollingSearchBox extends React.Component {
  static propTypes = {
    filterText: PropTypes.string.isRequired,
    onFilterTextChange: PropTypes.func.isRequired,
    spaceSearch: PropTypes.bool,
    children: PropTypes.element.isRequired
  }

  constructor () {
    super()
    this.state = {initialized: false}
  }

  onClick () {
    // Some browsers scroll at html level (chrome, firefox, ie11), some at body level (edge).
    const documentScroll = Math.max(document.documentElement.scrollTop, document.body.scrollTop)
    const elementTop = this._node.offsetTop
    if (documentScroll < elementTop) {
      this.scroll()
    }
  }

  scroll () {
    const elementTop = this._node.offsetTop
    animateScroll.scrollTo(elementTop, {duration: SCROLL_ANIMATION_DURATION})
  }

  onFilterTextChange (value) {
    this.props.onFilterTextChange(value)
    this.scroll()
  }

  setReference (node) {
    if (node) {
      this._node = node
      if (!this.state.initialized) {
        this.setState(Object.assign({}, this.state, {initialized: true}))
      }
    }
  }

  render () {
    return <div ref={node => this.setReference(node)}>
      <StickyHeadroom pinStart={this._node ? this._node.offsetTop : 0}
                      stickyNode={<SearchInput filterText={this.props.filterText}
                                               onFilterTextChange={value => this.onFilterTextChange(value)}
                                               onClickInput={() => this.onClick()}
                                               spaceSearch={this.props.spaceSearch} />}>
        {this.props.children}
      </StickyHeadroom>
    </div>
  }
}

export default ScrollingSearchBox
