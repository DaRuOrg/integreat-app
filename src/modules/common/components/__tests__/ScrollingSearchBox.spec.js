import React from 'react'
import { mount, shallow } from 'enzyme'
import ScrollingSearchBox from '../ScrollingSearchBox'
import SearchInput from '../SearchInput'
import { animateScroll } from 'react-scroll'

describe('ScrollingSearchBox', () => {
  const MockNode = () => <div />

  test('should render', () => {
    const component = (
      <ScrollingSearchBox filterText={'Test'}
                          onFilterTextChange={() => {}}>
        <MockNode />
      </ScrollingSearchBox>)
    expect(shallow(component)).toMatchSnapshot()
  })

  test('should space search', () => {
    expect(shallow(<ScrollingSearchBox filterText={'Test'}
                                       onFilterTextChange={() => {}}
                                       spaceSearch stickyTop={30}><MockNode /></ScrollingSearchBox>)).toMatchSnapshot()
  })

  test('should add stickyTop to containing Node', () => {
    const component = shallow(
      <ScrollingSearchBox filterText={'Test'}
                          onFilterTextChange={() => {}}
                          stickyTop={30}>
        <MockNode />
      </ScrollingSearchBox>)
    expect(component.find(MockNode).prop('stickyTop')).toEqual(75)
    expect(component.find('.searchBar').prop('style').top).toEqual('30px')
  })

  test('should pass onFilterTextChange and call scroll()', () => {
    const outerFilterTextChange = jest.fn()
    const component = mount(
      <ScrollingSearchBox filterText={'Test'}
                          onFilterTextChange={outerFilterTextChange}>
        <MockNode />
      </ScrollingSearchBox>)
    component.instance().scroll = jest.fn()

    component.find(SearchInput).prop('onFilterTextChange')('test')

    expect(outerFilterTextChange).toHaveBeenCalledWith('test')
    expect(component.instance().scroll).toHaveBeenCalled()
  })

  test('should set correct reference', () => {
    const component = mount(
      <ScrollingSearchBox filterText={'Test'} onFilterTextChange={() => {}}>
        <MockNode />
      </ScrollingSearchBox>)
    const node = component.instance()._node
    expect(node.className).toEqual('searchBar')
  })

  test('shouldnt call scroll() if user is already below searchInput onClick', () => {
    const component = mount(
      <ScrollingSearchBox filterText={'Test'} onFilterTextChange={() => {}}>
        <MockNode />
      </ScrollingSearchBox>)

    component.instance().scroll = jest.fn()
    component.instance()._node.getBoundingClientRect = () => ({top: 15})
    document.documentElement.scrollTop = 15

    component.find(SearchInput).prop('onClick')()
    expect(component.instance().scroll).toHaveLength(0)
  })

  test('should call scroll() if user is above searchInput onClick', () => {
    const component = mount(
      <ScrollingSearchBox filterText={'Test'} onFilterTextChange={() => {}}>
        <MockNode />
      </ScrollingSearchBox>)

    component.instance().scroll = jest.fn()
    component.instance()._node.getBoundingClientRect = () => ({top: 15})
    document.documentElement.scrollTop = 0

    component.find(SearchInput).prop('onClick')()
    expect(component.instance().scroll).toHaveBeenCalled()
  })

  test('should call animateScroll on scroll()', () => {
    const component = mount(
      <ScrollingSearchBox filterText={'Test'} onFilterTextChange={() => {}}>
        <MockNode />
      </ScrollingSearchBox>)

    component.instance()._node.getBoundingClientRect = () => ({top: 15})
    component.instance().scroll()

    expect(animateScroll.scrollTo).toHaveBeenCalledWith(15, {duration: 500})
  })
})
