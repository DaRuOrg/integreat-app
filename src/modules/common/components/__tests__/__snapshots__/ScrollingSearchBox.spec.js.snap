// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollingSearchBox should render 1`] = `
<React.Fragment>
  <div
    className="searchBar"
    style={
      Object {
        "top": "0px",
      }
    }
  >
    <Translate(SearchInput)
      filterText="Test"
      onClick={[Function]}
      onFilterTextChange={[Function]}
    />
  </div>
  <div
    className="searching"
  >
    <MockNode
      stickyTop={45}
    />
  </div>
</React.Fragment>
`;

exports[`ScrollingSearchBox should space search 1`] = `
<React.Fragment>
  <div
    className="searchBar"
    style={
      Object {
        "top": "30px",
      }
    }
  >
    <Translate(SearchInput)
      filterText="Test"
      onClick={[Function]}
      onFilterTextChange={[Function]}
      spaceSearch={true}
    />
  </div>
  <div
    className="searching"
  >
    <MockNode
      stickyTop={75}
    />
  </div>
</React.Fragment>
`;
